### DISCUSSION TITLE
Re-use ModelRelationManager form in ModelResource (or elsewhere)

### DISCUSSION THREAD

**Discussion by xavvey** [filamentphp/filament#3838] [Help] [answered]
This might be simpler than I think.. but I can't figure out how to re-use a form somewhere else...

I have a Client model and a ClientResource. A Client can have many Addresses so I made a AddressessRelationManager.
In this RelationManager I made the form to create a new Address.

I would like to re-use that same form for creating a new Address and add it to the form where a user can create a new Client. This, so that they don't first have to make a client and then after add the address, seeing as most of the time a Client will have only 1 Address.
If a Client has more Addresses then user can add them after creating the Client and first Address.

I've taken a look at the demo github page and found the AddressForm there, but I can't seem to find out how to use it...

Can someone tell me how to achieve this or point me in the right direction?

Thanks in advance!

**Comment by danharrin** [comment, marked as answer]
You could extract the form schema array into a getter method on the class, and then reuse it in both places?

```php
public static function getAddressFormSchema(): array
{
    return [
        TextInput::make(),
    ];
}
```

`ClassName::getAddressFormSchema()`

**Reply by xavvey** [reply]
Thanks for the fast reply!

Do you mean I should put the getter method inside the Address model class or in the RelationManager?

I tried it in de RelationManager as below but I get a `Filament\Forms\ComponentContainer::Filament\Forms\Concerns\{closure}(): Argument #1 ($component) must be of type Filament\Forms\Components\Component, array given` error...

```
`    public static function form(Form $form): Form
    {
        return $form
            ->schema(AddressessRelationManager::getAddressFormSchema());
    }

    public static function table(Table $table): Table
    {
        return $table
            ->columns([
            ])
            ->filters([
                //
            ])
            ->headerActions([
                CreateAction::make(),
            ])
            ->actions([
                EditAction::make(),
                DeleteAction::make(),
            ])
            ->bulkActions([
            ]);
    }

    public static function getAddressFormSchema(): array
    {
        return [
            TextColumn::make('street_name')
                ->label('STREET'),

            TextColumn::make('street_number')
                ->label('NUMBER'),

            TextColumn::make('postal_code')
                ->label('POSTAL CODE'),

            TextColumn::make('city')
                ->label('CITY'),

            BadgeColumn::make('address_type')
                ->label('TYPE')
                ->formatStateUsing(function ($record): string {
                    return $record->getFriendlyAddressTypeName();
                }),
        ];
    }` 
```

**Reply by ryangjchandler** [reply]
It's a static method so you can put it anywhere, but I'd probably put it on the resource itself.

**Reply by xavvey** [reply]
Thanks! I don't have an AddressResource so I placed it inside my AddressessRelationManager. Sorry if it was a 'dumb' question but thanks for taking your time and helping out!

**Reply by danharrin** [reply]
Youre trying to pass table Column objects not form
components which is causing the error

**Reply by xavvey** [reply]
I've noticed that error yeah...
